File: src/main/java/engine/rendering/instances/postprocessing/contrast/Contrast.java
package engine.rendering.instances.postprocessing.contrast;

import engine.rendering.RenderOutputData;
import engine.rendering.abstracted.postprocessing.AbstractPostProcessor;
import engine.utils.libraryBindings.opengl.shaders.RenderState;
import engine.utils.libraryBindings.opengl.shaders.UniformFloatProperty;
import engine.utils.libraryBindings.opengl.shaders.UniformTextureProperty;
import engine.utils.libraryBindings.opengl.textures.ITexture;

public class Contrast extends AbstractPostProcessor {

  private static final String FRAG_FILE = "/shaders/postprocessing/contrast/ContrastFragment.glsl";

  private float factor;

  public Contrast() throws Exception {
    this(1.4f); // Default factor
  }

  public Contrast(float factor) throws Exception {
    super(FRAG_FILE);
    this.factor = factor;
    initialize();
  }

  private void initialize() {
    getShadersProgram().addPerRenderUniform(new UniformTextureProperty<RenderOutputData>("textureSampler", 0) {
      @Override
      public ITexture getUniformValue(RenderState<RenderOutputData> state) {
        return state.getInstance().getColour();
      }
    });

    getShadersProgram().addPerRenderUniform(new UniformFloatProperty<RenderOutputData>("factor") {
      @Override
      public float getUniformValue(RenderState<RenderOutputData> state) {
        return factor;
      }
    });
  }
}

File: src/main/java/engine/rendering/instances/postprocessing/contrast/Contrast.kt
package engine.rendering.instances.postprocessing.contrast

import engine.rendering.RenderOutputData
import engine.rendering.abstracted.postprocessing.AbstractPostProcessor
import engine.utils.libraryBindings.opengl.shaders.RenderState
import engine.utils.libraryBindings.opengl.shaders.UniformFloatProperty
import engine.utils.libraryBindings.opengl.shaders.UniformTextureProperty
import engine.utils.libraryBindings.opengl.textures.ITexture
import engine.utils.libraryBindings.opengl.textures.Texture

class Contrast(var factor: Float = 1.4f) : AbstractPostProcessor(FRAG_FILE) {

    init {
        shadersProgram.addPerRenderUniform(object : UniformTextureProperty<RenderOutputData>("textureSampler", 0) {
            override fun getUniformValue(state: RenderState<RenderOutputData>): ITexture {
                return state.instance.colour
            }
        })
        shadersProgram.addPerRenderUniform(object : UniformFloatProperty<RenderOutputData>("factor") {
            override fun getUniformValue(state: RenderState<RenderOutputData>): Float {
                return factor
            }
        })
    }

    companion object {
        private const val FRAG_FILE = "/shaders/postprocessing/contrast/ContrastFragment.glsl"
    }

}
